src/app.rs

```diff
@@ -183,19 +305,114 @@ impl<'a, 'b, T: State + 'a> ApplicationBuilder<'a, 'b, T> {
-    pub fn register<C: Component>(mut self) -> Self {
+    /// Registers a component into the entity-component-system. This method
+    /// takes no options other then the Component's type which is defined
+    /// using a 'turbofish'. See the example for what this looks like.
+    ///
+    /// You must register a component's type before it can be used. If
+    /// code accesses a component that has not previously been registered
+    /// it will `panic`.
+    ///
+    /// # Type Parameters
+    ///
+    /// - `C`: The Component type that you are registering. This must
+    ///        implements the `Component` trait to be registered.
+    ///
+    /// # Returns
+    ///
+    /// This function returns ApplicationBuilder after it has modified it
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use amethyst::prelude::*;
+    /// use amethyst::ecs::{Component, HashMapStorage};
+    ///
+    /// struct NullState;
+    /// impl State for NullState {}
+    /// 
+    /// // define your custom type for the ECS
+    /// struct Velocity([f32; 3]);
+    ///
+    /// // The compiler must be told how to store every component, `Velocity`
+    /// // in this case. This is done via The `amethyst::ecs::Component` trait.
+    /// impl Component for Velocity {
+    ///     // To do this the Component trait has an associated type
+    ///     // which is used associate the type back to it's container's type.
+    ///     // There are a few common container's, VecStorage and HashMapStorage
+    ///     // are the most common used.
+    ///     type Storage = HashMapStorage<Velocity>;
+    /// }
+    ///
+    /// fn main() {
+    ///     // After creating a builder, we can add any number of components
+    ///     // using the register method.
+    ///     Application::build(NullState)
+    ///         .expect("Failed to initialize")
+    ///         .register::<Velocity>();
+    ///
+    /// }
+    /// ```
+    ///
+    pub fn register<C>(mut self) -> Self
+    where
+        C: Component,
+    {
         self.world.register::<C>();
         self
     }
 
-    /// Adds an ECS resource which can be accessed from systems.
-    pub fn add_resource<R>(mut self, res: R) -> Self
+    /// Adds the supplied ECS resource which can be accessed from game systems.
+    ///
+    /// Resources are common data that is shared with one or more game system.
+    /// 
+    /// # Parameters
+    /// `resource`: The initialized resource you wish to register
+    ///
+    /// # Type Parameters
+    ///
+    /// `R`: `resource` must implement the `Resource` trait. This trait will
+    ///      be automatically implemented if `Any` + `Send` + `Sync` traits
+    ///      exist for type `R`.
+    ///
+    /// # Returns
+    ///
+    /// This function returns ApplicationBuilder after it has modified it.
+    /// 
+    /// # Panics
+    ///
+    /// `add_resource` will panic if the resource with a matching type was
+    /// already registered.
```

Comment:
I'm not sure this is true. I believe specs::World will just overwrite the resource. It will panic if you try to register the same Component twice, but that's different.

Reply:
@Rhuagh  None of them causes a panic IIRC.

Reply:
Yes, see this line: https://github.com/slide-rs/specs/blob/master/src/world.rs#L606

Reply:
And here is the doc for `register`: https://github.com/slide-rs/specs/blob/master/src/world.rs#L588

Reply:
Ah, true, second Component register will just return straight away, while resources will be overwritten.

Reply:
Opps, my mistake.

I assumed that this panic would be hit https://github.com/slide-rs/shred/blob/master/src/res.rs#L253-L265

But https://github.com/slide-rs/specs/blob/master/src/world.rs#L593-L596 has a check that will avoid the panic.

I'll document that the resource being added will be silently dropped, since there be dragons.

Reply:
It's actually the case that if you call .add_resource(Camera::new(1)).add_resources(Camera::new(2)), the end result is that the 2nd one will be kept, and the first one thrown away. add_resource overwrites what was there before.